### PR TITLE
Slack bus relocation

### PR DESCRIPTION
**Pull request by geofjamg**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Bug fix

**What is the current behavior?**

When a contingency lost the slack bus we skip the contingency.

**What is the new behavior (if this is a feature change)?**
We now relocate the slack bus to another valid bus for the contingency and then restore the original slack bus for the next contingency. 

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

**Other information**:

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/network/impl/PropagatedContingency.java [hunk 1]
```diff
@@ -390,7 +397,8 @@ private Map<LfBranch, DisabledBranchStatus> findBranchToOpenDirectlyImpactedByCo
     record ContingencyConnectivityLossImpact(boolean ok, int createdSynchronousComponents, Set<LfBus> busesToLost, Set<LfHvdc> hvdcsWithoutPower) {
     }
 
-    private ContingencyConnectivityLossImpact findBusesAndBranchesImpactedBecauseOfConnectivityLoss(LfNetwork network, Map<LfBranch, DisabledBranchStatus> branchesToOpen) {
+    private ContingencyConnectivityLossImpact findBusesAndBranchesImpactedBecauseOfConnectivityLoss(LfNetwork network, Map<LfBranch, DisabledBranchStatus> branchesToOpen,
+                                                                                                    Set<LfBus> initialBusIdsToLose) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/PropagatedContingency.java:401]: You can remove initialBusIdsToLose from attributes

File: src/main/java/com/powsybl/openloadflow/network/LfNetwork.java [hunk 2]
```diff
@@ -224,6 +237,9 @@ public void updateSlackBusesAndReferenceBus() {
                 LOGGER.info("Network {}, reference generator is {}", this, referenceGenerator.getId());
                 referenceGenerator.setReference(true);
             }
+            if (connectivity != null) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/LfNetwork.java:240]: We have another place in the code with `connectivity.setMainComponentVertex(getSlackBuses().get(0));`, is it normal?

Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/LfNetwork.java:240]: ok, understand.

**Review by annetill — APPROVED** [review]

File: src/main/java/com/powsybl/openloadflow/network/impl/PropagatedContingency.java [hunk 3]
```diff
@@ -399,12 +406,15 @@ private ContingencyConnectivityLossImpact findBusesAndBranchesImpactedBecauseOfC
                     .filter(LfBranch::isConnectedAtBothSides)
                     .forEach(connectivity::removeEdge);
 
-            if (connectivity.getConnectedComponent(network.getSlackBus()).size() == 1) {
-                // FIXME
-                // If a contingency leads to an isolated slack bus, this bus is considered as the main component.
-                // In that case, we have an issue with a different number of variables and equations.
-                LOGGER.error("Contingency '{}' leads to an isolated slack bus: not supported", contingency.getId());
-                return new ContingencyConnectivityLossImpact(false, 0, Collections.emptySet(), Collections.emptySet());
+            if (isSlackBusIsolated(connectivity, network.getSlackBus())) {
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/network/impl/PropagatedContingency.java:409]:
```suggestion
            if (network.getSlackBuses().stream().anyMatch(slackBus -> isSlackBusIsolated(connectivity, slackBus))) {
```

fixes a modified version of `testBusBarSectionContingencyWithSlackBusRelocation()` with following parameters:
```java
lfParameters.getExtension(OpenLoadFlowParameters.class).setSlackBusesIds(List.of("VL1_1", "VL1_0")).setMaxSlackBusCount(2);
```

but remains an issue to solve, i am getting `Expected to have same number of equations (9) and variables (10)` with
```java
lfParameters.getExtension(OpenLoadFlowParameters.class).setSlackBusesIds(List.of("VL1_0")).setMaxSlackBusCount(2);
```

for N situ:
```
INFO  14:26:59.050 c.p.o.n.LfNetwork    | Network {CC0 SC0}, slack buses are [VL1_0] (method='Parameter bus')
INFO  14:26:59.051 c.p.o.n.LfNetwork    | Network {CC0 SC0}, reference bus is VL1_0 (method='First slack')
```

and CTG situation:
```
INFO  14:26:59.216 c.p.o.n.LfNetwork    | Network {CC0 SC0}, slack buses are [VL1_1, VL2_0] (method='Most meshed bus')
TRACE 14:26:59.217 c.p.o.n.LfNetworkListenerTracer | onSlackBusChange(bus=VL1_1, slack=true)
TRACE 14:26:59.217 c.p.o.n.LfNetworkListenerTracer | onSlackBusChange(bus=VL2_0, slack=true)
INFO  14:26:59.217 c.p.o.n.LfNetwork    | Network {CC0 SC0}, reference bus is VL1_1 (method='First slack')
TRACE 14:26:59.217 c.p.o.n.LfNetworkListenerTracer | onReferenceBusChange(bus=VL1_1, reference=true)
...

java.util.concurrent.CompletionException: com.powsybl.commons.PowsyblException: Expected to have same number of equations (9) and variables (10)
```

Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/network/impl/PropagatedContingency.java:409]: probably a weird edge case where with provided settings we get more slack buses in N-1 situation compared to N situation

**Review by jeandemanged — APPROVED** [review]
support of multiple slack buses will be revisited in other PRs

**Comment by annetill** [comment]
> support of multiple slack buses will be revisited in other PRs

And the possibility to have for a security analysis a slack bus different that the reference bus. And of course the support of the lose of the reference bus or the slack bus.

**Comment by geofjamg** [comment]
@jeandemanged we decided to just no support AS with multiple slack buses and write a log to warm that we have forced the slack bus count to 1. It would need a lot a dev to support it as it would require to correctly update the active power distribution equation between slack each time a slack bus is disabled or enabled. I probably does not worth to do it as mutliple slacks is very rarely useful.

**Pull request merged**